components/pd_client/src/util.rs

```diff
@@ -780,3 +783,191 @@ pub fn check_resp_header(header: &ResponseHeader) -> Result<()> {
+        let end = &incoming[new_idx + 1];
+        if let Some((start_idx, end_idx)) = find_overlay_ranges((start.as_ref(), end.as_ref()), cur)
+        {
+            for cur_idx in start_idx..=end_idx {
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, read_bytes);
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, write_bytes);
+
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, read_qps);
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, write_qps);
+
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, read_keys);
+                stats_add!(cur_stats, cur_idx, delta_stats, new_idx, write_keys);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{merge_bucket_stats, util::find_bucket_index};
+    use kvproto::metapb::BucketStats;
+
+    #[test]
+    fn test_merge_bucket_stats() {
+        let cases = [
+            (
+                (vec![b"k1", b"k3", b"k5", b"k7", b"k9"], vec![1, 1, 1, 1]),
+                (vec![b"k1", b"k3", b"k5", b"k7", b"k9"], vec![1, 1, 1, 1]),
+                vec![2, 2, 2, 2],
+            ),
+            (
+                (vec![b"k1", b"k3", b"k5", b"k7", b"k9"], vec![1, 1, 1, 1]),
+                (vec![b"k0", b"k6", b"k8"], vec![1, 1]),
+                vec![2, 2, 3, 2],
+            ),
+            (
+                (vec![b"k0", b"k6", b"k8"], vec![1, 1]),
+                (
+                    vec![b"k1", b"k3", b"k5", b"k7", b"k9", b"ka"],
+                    vec![1, 1, 1, 1, 1],
+                ),
+                vec![4, 3],
+            ),
+            (
+                (vec![b"k4", b"k6", b"kb"], vec![1, 1]),
+                (
+                    vec![b"k1", b"k3", b"k5", b"k7", b"k9", b"ka"],
+                    vec![1, 1, 1, 1, 1],
+                ),
+                vec![3, 4],
+            ),
+        ];
+        for (current, incoming, expected) in cases {
+            let cur_keys = current.0;
+            let incoming_keys = incoming.0;
+            let mut cur_stats = BucketStats::default();
+            cur_stats.set_read_qps(current.1);
+            let mut incoming_stats = BucketStats::default();
+            incoming_stats.set_read_qps(incoming.1);
+            merge_bucket_stats(&cur_keys, &mut cur_stats, &incoming_keys, &incoming_stats);
+            assert_eq!(cur_stats.get_read_qps(), expected);
+        }
+    }
+
+    #[test]
+    fn test_find_bucket_index() {
+        let keys = vec![
+            b"k1".to_vec(),
+            b"k3".to_vec(),
+            b"k5".to_vec(),
+            b"k7".to_vec(),
+        ];
+        assert_eq!(find_bucket_index(b"k1", &keys), Some(0));
+        assert_eq!(find_bucket_index(b"k5", &keys), Some(2));
+        assert_eq!(find_bucket_index(b"k2", &keys), Some(0));
+        assert_eq!(find_bucket_index(b"k6", &keys), Some(2));
+        assert_eq!(find_bucket_index(b"k7", &keys), Some(2));
```

Comment:
```suggestion
        assert_eq!(find_bucket_index(b"k7", &keys), None);
```

Reply:
The last key of buckets is the end of the region, so the bucket range will be `[key[i], key[i+1])`. cc @bufferflies